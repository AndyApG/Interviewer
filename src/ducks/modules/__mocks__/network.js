/* eslint-env jest */

import uuidv4 from '../../../utils/uuid';

const { actionTypes, entityPrimaryKeyProperty } = jest.requireActual('../network');

const reducer = jest.fn(() => ({
  nodes: [],
  edges: [],
  ego: {
    [entityPrimaryKeyProperty]: uuidv4(),
  },
}));

export default reducer;

export {
  actionTypes,
};
