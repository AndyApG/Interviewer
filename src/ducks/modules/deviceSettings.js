import deviceDescription from '../../utils/DeviceInfo';

const SET_DESCRIPTION = 'SETTINGS/SET_DESCRIPTION';
const SET_INTERFACE_SCALE = 'SETTINGS/SET_INTERFACE_SCALE';
const TOGGLE_SETTING = 'SETTINGS/TOGGLE_SETTING';

const initialState = {
  description: deviceDescription(),
  useFullScreenForms: window.matchMedia('screen and (min-device-aspect-ratio: 8/5), (min-device-height: 1800px)').matches,
  useDynamicScaling: true,
  interfaceScale: 100,
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SET_DESCRIPTION:
      return {
        ...state,
        description: action.description,
      };
    case SET_INTERFACE_SCALE:
      return {
        ...state,
        interfaceScale: action.scale,
      };
    case TOGGLE_SETTING:
      return {
        ...state,
        [action.item]: !state[action.item],
      };
    default:
      return state;
  }
}

const setDescription = description => ({
  type: SET_DESCRIPTION,
  description,
});

const setInterfaceScale = scale => ({
  type: SET_INTERFACE_SCALE,
  scale,
});

const toggleSetting = item => ({
  type: TOGGLE_SETTING,
  item,
});

const actionCreators = {
  setDescription,
  setInterfaceScale,
  toggleSetting,
};

const actionTypes = {
  SET_DESCRIPTION,
  SET_INTERFACE_SCALE,
  TOGGLE_SETTING,
};

export {
  actionCreators,
  actionTypes,
};
