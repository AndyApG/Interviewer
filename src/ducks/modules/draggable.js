const DRAG_START = 'DRAG_START';
const DRAG_STOP = 'DRAG_STOP';

const initialState = {
  isDragging: false,
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
      case DRAG_START:
        return {
          ...state,
          isDragging: true,
        }
        case DRAG_STOP:
          return {
            ...state,
            isDragging: false,
          }
    default:
      return state;
  }
};

function dragStart() {
  return {
    type: DRAG_START,
  }
};

function dragStop() {
  return {
    type: DRAG_STOP,
  }
};

const actionCreators = {
  dragStart,
  dragStop,
};

const actionTypes = {
  DRAG_START,
  DRAG_STOP,
};

export {
  actionCreators,
  actionTypes,
};
