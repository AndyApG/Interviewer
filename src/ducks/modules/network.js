import { reject, findIndex, isMatch, omit } from 'lodash';

import uuidv4 from '../../utils/uuid';

// Property name for the primary key for nodes
export const NodePrimaryKeyProperty = '_uid';

// Property name for node "model" properties
export const NodeAttributesProperty = 'attributes';

export const ADD_NODES = 'ADD_NODES';
export const REMOVE_NODE = 'REMOVE_NODE';
export const UPDATE_NODE = 'UPDATE_NODE';
export const TOGGLE_NODE_ATTRIBUTES = 'TOGGLE_NODE_ATTRIBUTES';
export const ADD_EDGE = 'ADD_EDGE';
export const TOGGLE_EDGE = 'TOGGLE_EDGE';
export const REMOVE_EDGE = 'REMOVE_EDGE';
export const SET_EGO = 'SET_EGO';
export const UNSET_EGO = 'UNSET_EGO';


// Initial network model structure
const initialState = {
  ego: {},
  nodes: [],
  edges: [],
};

function flipEdge(edge) {
  return { from: edge.to, to: edge.from, type: edge.type };
}

function edgeExists(edges, edge) {
  return (
    findIndex(edges, edge) !== -1 ||
    findIndex(edges, flipEdge(edge)) !== -1
  );
}

/**
 * existingNodes - Existing network.nodes
 * netNodes - nodes to be added to the network
 * additionalAttributes - node model properties
*/
function getNodesWithBatchAdd(existingNodes, newNodes, nodeAttributeData) {
  // Create a function to create a UUID and merge node attributes
  const withModelandAttributeData = newNode => ({
    [NodePrimaryKeyProperty]: uuidv4(),
    ...newNode, // second to allow existing UUID to be overwritten
    [NodeAttributesProperty]: {
      ...newNode[NodeAttributesProperty],
      ...nodeAttributeData,
    },
  });

  return existingNodes.concat(newNodes.map(withModelandAttributeData));
}

/**
 * @param {array} nodes - an array of objects representing nodes
 * @param {object} updatedNode - object representing the node to be updated and its new properties
 * @param {boolean} full - if the entire node should be overwritten
 */
function getUpdatedNodes(nodes, updatedNode, nodeAttributeData) {
  // Iterate over the nodes list
  const updatedNodes = nodes.map((node) => {
    // Skip nodes where the primary key doesn't match
    if (node[NodePrimaryKeyProperty] !== updatedNode[NodePrimaryKeyProperty]) { return node; }
    console.log('getUpdatedNodes');
    // if we have an attributes payload, merge with any existing attributes
    if (nodeAttributeData) {
      console.log('nodeAttributedata found');
      console.log(updatedNode, nodeAttributeData);
      return {
        ...node,
        ...updatedNode,
        [NodeAttributesProperty]: {
          ...node[NodeAttributesProperty],
          ...nodeAttributeData,
        },
        [NodePrimaryKeyProperty]: node[NodePrimaryKeyProperty],
      };
    }

    // If no attribute payload, just merge the new node data with the existing
    return {
      ...node,
      ...updatedNode,
      [NodePrimaryKeyProperty]: node[NodePrimaryKeyProperty],
    };
  });

  // Return the modified array of nodes
  return updatedNodes;
}

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case ADD_NODES: {
      return {
        ...state,
        nodes: getNodesWithBatchAdd(state.nodes, action.nodes, action.additionalAttributes),
      };
    }
    /**
     * TOGGLE_NODE_ATTRIBUTES
     * Used to toggle the value of a boolean variable between states when the node is tapped.
     */
    case TOGGLE_NODE_ATTRIBUTES: {
      // attributes = object containing node properties, minus _uid
      const attributes = omit(action.attributes, [NodePrimaryKeyProperty]);

      // Map over the nodes
      const updatedNodes = state.nodes.map((node) => {
        // Skip nodes with different primary keys
        if (node[NodePrimaryKeyProperty] !== action[NodePrimaryKeyProperty]) { return node; }

        //
        if (isMatch(node, attributes)) {
          return omit(node, Object.getOwnPropertyNames(attributes));
        }

        return {
          ...node,
          ...action.attributes,
        };
      });

      return {
        ...state,
        nodes: updatedNodes,
      };
    }
    case UPDATE_NODE: {
      return {
        ...state,
        nodes: getUpdatedNodes(state.nodes, action.node, action.additionalAttributes),
      };
    }
    case REMOVE_NODE: {
      const removeNodePrimaryKeyProperty = action[NodePrimaryKeyProperty];
      return {
        ...state,
        nodes: reject(state.nodes, node =>
          node[NodePrimaryKeyProperty] === removeNodePrimaryKeyProperty),
        edges: reject(state.edges, edge =>
          edge.from === removeNodePrimaryKeyProperty || edge.to === removeNodePrimaryKeyProperty),
      };
    }
    case ADD_EDGE:
      if (edgeExists(state.edges, action.edge)) { return state; }
      return {
        ...state,
        edges: [...state.edges, action.edge],
      };
    case TOGGLE_EDGE:
      // remove edge if it exists, add it if it doesn't
      if (edgeExists(state.edges, action.edge)) {
        return {
          ...state,
          edges: reject(reject(state.edges, action.edge), flipEdge(action.edge)),
        };
      }
      return {
        ...state,
        edges: [...state.edges, action.edge],
      };
    case REMOVE_EDGE:
      if (edgeExists(state.edges, action.edge)) {
        return {
          ...state,
          edges: reject(reject(state.edges, action.edge), flipEdge(action.edge)),
        };
      }
      return state;
    default:
      return state;
  }
}

/**
 * @namespace NetworkActionCreators
 */
const actionCreators = {};

const actionTypes = {
  ADD_NODES,
  UPDATE_NODE,
  TOGGLE_NODE_ATTRIBUTES,
  REMOVE_NODE,
  ADD_EDGE,
  TOGGLE_EDGE,
  REMOVE_EDGE,
  SET_EGO,
  UNSET_EGO,
};

export {
  actionCreators,
  actionTypes,
};
