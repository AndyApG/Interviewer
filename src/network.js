let namesList = ['Joshua', 'Bernie', 'Michelle', 'Gregory', 'Patrick', 'Barney', 'Jonathon','Myles','Alethia','Tammera','Veola','Meredith','Renee','Grisel','Celestina','Fausto','Eliana','Raymundo','Lyle','Carry','Kittie','Melonie','Elke','Mattie','Kieth','Lourie','Marcie','Trinity','Librada','Lloyd','Pearlie','Velvet','Stephan','Hildegard','Winfred','Tempie','Maybelle','Melynda','Tiera','Lisbeth','Kiera','Gaye','Edra','Karissa','Manda','Ethelene','Michelle','Pamella','Jospeh','Tonette','Maren','Aundrea','Madelene','Epifania','Olive'];

module.exports = class Network {

  constructor() {
      this._edges = [];
      this._nodes = [];
      this.reserved_ids = [];
  }

  addNode(properties) {

    if (!this.egoExists()) {
      throw new Error('You must add an Ego before attempting to add other nodes.');
    }

    let newNodeID = 0;

    while (this.getNode(newNodeID) !== false || this.reserved_ids.indexOf(newNodeID) !== -1) {
      newNodeID++;
    }

    let nodeProperties = {
      id: newNodeID
    };

    Object.assign(nodeProperties, properties); // ES6 approach to extending objects.

    this.nodes.push(nodeProperties);
    this.reserved_ids.push(newNodeID);
  }

  egoExists() {
    return true;
  }

  resetNetwork() {
    this.nodes = [];
    this.edges = [];
  }

  getNode(id) {
    // Ensure that ID is always treated as int for === comparisons to work reliably.
    id = parseInt(id);

    if (id === undefined) { return false; }
    for (var i = 0;i<this._nodes.length; i++) {
      if (this._nodes[i].id === id) {return this._nodes[i]; }
    }
    return false;

  }

  get nodes() {
    return this._nodes;
  }

};
