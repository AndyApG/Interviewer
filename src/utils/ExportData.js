import { findKey, forInRight, isNil, join } from 'lodash';

import saveFile from './SaveFile';
import { NodePrimaryKeyProperty } from '../ducks/modules/network';

const setUpXml = () => {
  const xmlDoc = '<?xml version="1.0" encoding="UTF-8"?>\n' +
    '<graphml xmlns="http://graphml.graphdrawing.org/xmlns"\n' +
    'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"\n' +
    'xsi:schemaLocation="http://graphml.graphdrawing.org/xmlns\n' +
    'http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd">\n' +
    '  <graph edgedefault="undirected">\n' +
    ' </graph>\n' +
    ' </graphml>\n';

  let xml = '';
  if (window.DOMParser) { // Standard
    xml = (new DOMParser()).parseFromString(xmlDoc, 'text/xml');
  }
  return xml;
};

const getVariableInfo = (variableRegistry, type, element, key) => (
  variableRegistry[type] && variableRegistry[type][element.type] &&
  variableRegistry[type][element.type].variables &&
  variableRegistry[type][element.type].variables[key]
);

const isVariableRegistryExists = (variableRegistry, type, element, key) => {
  const knownTypes = ['boolean', 'text', 'number', 'datetime', 'categorical', 'ordinal', 'layout',
    'location'];
  const variableInfo = getVariableInfo(variableRegistry, type, element, key);
  return variableInfo && variableInfo.type && knownTypes.includes(variableInfo.type);
};

const getTypeFromVariableRegistry = (variableRegistry, type, element, key) => {
  const variableInfo = getVariableInfo(variableRegistry, type, element, key);
  return variableInfo && variableInfo.type;
};

const getTypeForKey = (data, key) => (
  data.reduce((result, value) => {
    if (isNil(value[key])) return result;
    let currentType = typeof value[key];
    if (currentType === 'number') {
      currentType = Number.isInteger(value[key]) ? 'integer' : 'double';
      if (result && currentType !== result) return 'double';
    }
    if (String(Number.parseInt(value[key], 10)) === value[key]) {
      currentType = 'integer';
      if (result === 'double') return 'double';
    } else if (String(Number.parseFloat(value[key], 10)) === value[key]) {
      currentType = 'double';
      if (result === 'integer') return 'double';
    }
    if (isNil(currentType)) return result;
    if (currentType === result || result === '') return currentType;
    return 'string';
  }, ''));

const generateKeys = (graph, graphML, elements, type, excludeList, variableRegistry,
  layoutVariable) => {
  // generate keys for attributes
  const missingVariables = [];
  const done = [];

  // add keys for gephi positions
  if (layoutVariable) {
    const xElement = document.createElementNS(graphML.namespaceURI, 'key');
    xElement.setAttribute('id', 'x');
    xElement.setAttribute('attr.name', 'x');
    xElement.setAttribute('attr.type', 'double');
    xElement.setAttribute('for', type);
    graphML.insertBefore(xElement, graph);
    const yElement = document.createElementNS(graphML.namespaceURI, 'key');
    yElement.setAttribute('id', 'y');
    yElement.setAttribute('attr.name', 'y');
    yElement.setAttribute('attr.type', 'double');
    yElement.setAttribute('for', type);
    graphML.insertBefore(yElement, graph);
  }

  elements.forEach((element) => {
    Object.keys(element).forEach((key) => {
      if (done.indexOf(key) === -1 && !excludeList.includes(key)) {
        const keyElement = document.createElementNS(graphML.namespaceURI, 'key');
        keyElement.setAttribute('id', key);
        keyElement.setAttribute('attr.name', key);

        if (!isVariableRegistryExists(variableRegistry, type, element, key)) {
          missingVariables.push(`"${key}" in ${type}.${element.type}`);
        }

        switch (getTypeFromVariableRegistry(variableRegistry, type, element, key)) {
          case 'boolean':
            keyElement.setAttribute('attr.type', 'boolean');
            break;
          case 'integer':
            keyElement.setAttribute('attr.type', 'integer');
            break;
          case 'double':
            keyElement.setAttribute('attr.type', 'double');
            break;
          case 'ordinal':
          case 'number': {
            const keyType = getTypeForKey(elements, key);
            keyElement.setAttribute('attr.type', keyType);
            break;
          }
          case 'layout': {
            // special handling for locations
            keyElement.setAttribute('attr.name', `${key}Y`);
            keyElement.setAttribute('id', `${key}Y`);
            keyElement.setAttribute('attr.type', 'double');
            const keyElement2 = document.createElementNS(graphML.namespaceURI, 'key');
            keyElement2.setAttribute('id', `${key}X`);
            keyElement2.setAttribute('attr.name', `${key}X`);
            keyElement2.setAttribute('attr.type', 'double');
            keyElement2.setAttribute('for', type);
            graphML.insertBefore(keyElement2, graph);
            break;
          }
          case 'text':
          case 'datetime':
          case 'categorical':
          case 'location':
          default:
            keyElement.setAttribute('attr.type', 'string');
        }
        keyElement.setAttribute('for', type);
        graphML.insertBefore(keyElement, graph);
        done.push(key);
      }
    });
  });
  return missingVariables;
};

const getDataElement = (uri, key, text) => {
  const data = document.createElementNS(uri, 'data');
  data.setAttribute('key', key);
  data.appendChild(document.createTextNode(text));
  return data;
};

const addElements = (
  graph,
  uri,
  dataList,
  type,
  excludeList,
  variableRegistry,
  layoutVariable,
  extra = false) => {
  dataList.forEach((dataElement, index) => {
    const domElement = document.createElementNS(uri, type);
    if (dataElement[NodePrimaryKeyProperty]) {
      domElement.setAttribute('id', dataElement[NodePrimaryKeyProperty]);
    } else {
      domElement.setAttribute('id', index);
    }
    if (extra) domElement.setAttribute('source', dataElement.from);
    if (extra) domElement.setAttribute('target', dataElement.to);
    graph.appendChild(domElement);

    Object.keys(dataElement).forEach((key) => {
      if (!excludeList.includes(key)) {
        if (typeof dataElement[key] !== 'object') {
          domElement.appendChild(getDataElement(uri, key, dataElement[key]));
        } else if (getTypeFromVariableRegistry(variableRegistry, type, dataElement, key) === 'layout') {
          domElement.appendChild(getDataElement(uri, `${key}X`, dataElement[key].x));
          domElement.appendChild(getDataElement(uri, `${key}Y`, dataElement[key].y));
        } else {
          domElement.appendChild(getDataElement(uri, key, JSON.stringify(dataElement[key])));
        }
      }
    });

    // add positions for gephi
    if (layoutVariable && dataElement[layoutVariable]) {
      domElement.appendChild(getDataElement(uri, 'x', dataElement[layoutVariable].x * window.innerWidth));
      domElement.appendChild(getDataElement(uri, 'y', (1.0 - dataElement[layoutVariable].y) * window.innerHeight));
    }
  });
};

const xmlToString = (xmlData) => {
  let xmlString;
  if (window.ActiveXObject) { // IE
    xmlString = xmlData.xml;
  } else { // code for Mozilla, Firefox, Opera, etc.
    xmlString = (new window.XMLSerializer()).serializeToString(xmlData);
  }
  return xmlString;
};

const createGraphML = (networkData, variableRegistry, openErrorDialog) => {
  // default graph structure
  const xml = setUpXml();
  const graph = xml.getElementsByTagName('graph')[0];
  const graphML = xml.getElementsByTagName('graphml')[0];

  // find the first variable of type layout
  let layoutVariable;
  forInRight(variableRegistry.node, (value) => {
    layoutVariable = findKey(value.variables, { type: 'layout' });
  });

  // generate keys for attributes
  let missingVariables = generateKeys(graph, graphML, networkData.nodes, 'node', [NodePrimaryKeyProperty], variableRegistry, layoutVariable);
  missingVariables = missingVariables.concat(generateKeys(graph, graphML, networkData.edges, 'edge', ['from', 'to'], variableRegistry));
  if (missingVariables.length > 0) {
    // hard fail if checking the registry fails
    // remove this to fall back to using "text" for unknowns
    openErrorDialog(`The variable registry seems to be missing "type" of: ${join(missingVariables, ', ')}.`);
    return null;
  }

  // add nodes and edges to graph
  addElements(graph, graphML.namespaceURI, networkData.nodes, 'node', [NodePrimaryKeyProperty], variableRegistry, layoutVariable);
  addElements(graph, graphML.namespaceURI, networkData.edges, 'edge', ['from', 'to'], variableRegistry, null, true);

  return saveFile(xmlToString(xml), openErrorDialog, 'graphml', ['graphml'], 'networkcanvas.graphml', 'text/xml',
    { message: 'Your network canvas graphml file.', subject: 'network canvas export' });
};

export default createGraphML;
