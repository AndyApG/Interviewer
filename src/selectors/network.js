import { findKey } from 'lodash';
import { getActiveSession, getCaseId } from './session';
import { createDeepEqualSelector } from './utils';
import { getProtocolCodebook } from './protocol';
import { asExportableNetwork, asWorkerAgentNetwork } from '../utils/networkFormat';
import { getEntityAttributes } from '../ducks/modules/network';

export const getNetwork = createDeepEqualSelector(
  (state, props) => getActiveSession(state, props),
  session => (session && session.network) || { nodes: [], edges: [] },
);

export const getNetworkNodes = createDeepEqualSelector(
  getNetwork,
  network => network.nodes,
);

export const getNetworkEgo = createDeepEqualSelector(
  getNetwork,
  network => network.ego,
);

export const getNetworkEdges = createDeepEqualSelector(
  getNetwork,
  network => network.edges,
);

export const getExportableNetwork = createDeepEqualSelector(
  getNetwork,
  (state, props) => getProtocolCodebook(state, props),
  (state, props) => getCaseId(state, props),
  (network, registry, _caseID) => asExportableNetwork(network, registry, { _caseID }),
);

export const getWorkerNetwork = createDeepEqualSelector(
  getNetwork,
  (state, props) => getProtocolCodebook(state, props),
  (network, registry) => asWorkerAgentNetwork(network, registry),
);

// The user-defined name of a node type; e.g. `codebook.node[uuid].name == 'person'`
export const makeGetNodeTypeDefinition = () => createDeepEqualSelector(
  (state, props) => getProtocolCodebook(state, props),
  (state, props) =>
    (props && props.type) ||
    (props && props.stage && props.stage.subject && props.stage.subject.type) ||
    (state && state.type),
  (codebook, nodeType) => {
    const nodeInfo = codebook && codebook.node;
    return nodeInfo && nodeInfo[nodeType];
  },
);

// See: https://github.com/codaco/Network-Canvas/wiki/Node-Labeling
const labelLogic = (codebookForNodeType, nodeAttributes) => {
  // In the codebook for the stage's subject, look for the displayVariable property
  // and try to retrieve this value as a key in the node's attributes.
  const displayVariable = codebookForNodeType && codebookForNodeType.displayVariable;
  if (displayVariable && nodeAttributes[displayVariable]) {
    return nodeAttributes[displayVariable];
  }

  // In the codebook for the stage's subject, look for a variable with a name
  // property of "name", and try to retrieve this value by key in the node's
  // attributes
  const variableCalledName = codebookForNodeType && codebookForNodeType.variables && findKey(codebookForNodeType.variables, ['name', 'name']);
  if (variableCalledName && nodeAttributes[variableCalledName]) {
    return nodeAttributes[variableCalledName];
  }

  // look for a property on the node with a key of ‘displayVariable’, and try to
  // retrieve this value as a key in the node's attributes.
  const nodeDisplayVariable = nodeAttributes.displayVariable;
  if (nodeDisplayVariable && nodeAttributes[nodeDisplayVariable]) {
    return nodeAttributes[nodeDisplayVariable];
  }

  // look for a property on the node with a key of ‘name’, and try to retrieve this
  // value as a key in the node's attributes.
  const nodeVariableCalledName = findKey(nodeAttributes, ['name', 'name']);
  if (nodeVariableCalledName) {
    return nodeAttributes[nodeVariableCalledName];
  }

  return 'No label';
};

// Gets the node label variable and returns its value, or "No label".
// See: https://github.com/codaco/Network-Canvas/wiki/Node-Labeling
export const makeGetNodeLabel = () => createDeepEqualSelector(
  (state, props) => {
    const getNodeTypeDefinition = makeGetNodeTypeDefinition(state, props);
    return getNodeTypeDefinition(state, props);
  },
  nodeTypeDefinition => node => labelLogic(nodeTypeDefinition, getEntityAttributes(node)),
);

export const makeGetNodeColor = () => createDeepEqualSelector(
  getProtocolCodebook,
  (_, props) => props.type,
  (codebook, nodeType) => {
    const nodeInfo = codebook.node;
    return (nodeInfo && nodeInfo[nodeType] && nodeInfo[nodeType].color) || 'node-color-seq-1';
  },
);

export const makeGetEdgeLabel = () => createDeepEqualSelector(
  getProtocolCodebook,
  (_, props) => props.type,
  (codebook, edgeType) => {
    const edgeInfo = codebook.edge;
    return (edgeInfo && edgeInfo[edgeType] && edgeInfo[edgeType].label) || '';
  },
);

export const makeGetEdgeColor = () => createDeepEqualSelector(
  getProtocolCodebook,
  (_, props) => props.type,
  (codebook, edgeType) => {
    const edgeInfo = codebook.edge;
    return (edgeInfo && edgeInfo[edgeType] && edgeInfo[edgeType].color) || 'edge-color-seq-1';
  },
);

export const makeGetNodeAttributeLabel = () => createDeepEqualSelector(
  getProtocolCodebook,
  (_, props) => props.subject.type,
  (_, props) => props.variableId,
  (codebook, nodeType, variableId) => {
    const nodeInfo = codebook.node;
    const variables = (nodeInfo && nodeInfo[nodeType] && nodeInfo[nodeType].variables) || {};
    return (variables && variables[variableId] && variables[variableId].label) || [];
  },
);

export const makeGetCategoricalOptions = () => createDeepEqualSelector(
  (state, props) => getProtocolCodebook(state, props),
  (_, props) => props.subject.type,
  (_, props) => props.variableId,
  (codebook, nodeType, variableId) => {
    const nodeInfo = codebook.node;
    const variables = (nodeInfo && nodeInfo[nodeType] && nodeInfo[nodeType].variables) || {};
    return (variables && variables[variableId] && variables[variableId].options) || [];
  },
);
