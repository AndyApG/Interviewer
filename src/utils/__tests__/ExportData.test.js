/* eslint-env jest */
import saveFile from '../SaveFile';
import ExportData from '../ExportData';
import { NodePK } from '../../ducks/modules/network';

function mockSerializeToString() {
  return { serializeToString: xmlData => xmlData.documentElement.outerHTML };
}
window.XMLSerializer = window.XMLSerializer || mockSerializeToString;

jest.mock('../SaveFile');
saveFile.mockImplementation(data => data);

const variableRegistry = {
  node: {
    person: {
      variables: {
        type: { type: 'text' },
        name: { type: 'text' },
        aString: { type: 'text' },
        aNumber: { type: 'number' },
        bNumber: { type: 'number' },
        cNumber: { type: 'number' },
        aBoolean: { type: 'boolean' },
        aDatetime: { type: 'datetime' },
        aCategorical: { type: 'categorical' },
        aOrdinal: { type: 'ordinal' },
        aLayout: { type: 'layout' },
        aLocation: { type: 'location' },
      },
    },
  },
  edge: {
    friend: {
      variables: {
        type: { type: 'text' },
        connected: { type: 'boolean' },
      },
    },
  },
};

const sessionA = {
  network: {
    edges: [
      { type: 'friend', to: 1, from: 2, connected: true },
    ],
    edgo: {},
    nodes: [
      { [NodePK]: 1,
        type: 'person',
        name: 'soAndSo',
        aString: 'content',
        aNumber: 123,
        bNumber: 0.33,
        cNumber: 3,
        aBoolean: true,
        aDatetime: 1529349451847,
        aCategorical: 24,
        aOrdinal: 1,
        aLayout: { x: 0.4134, y: 0.2356 },
        aLocation: { latitude: 41.799756, longitude: -87.66443 },
      },
      { [NodePK]: 2,
        type: 'person',
        name: 'whoDunnit',
        aString: 'Another Content',
        aNumber: 456,
        bNumber: 1.8,
        cNumber: 33,
        aBoolean: false,
        aDatetime: 1529343481847,
        aCategorical: 77,
        aOrdinal: 2,
        aLayout: { x: 0.3434, y: 0.3156 },
        aLocation: { latitude: 44.9756, longitude: 18.443 },
      },
      { [NodePK]: 3,
        type: 'person',
        name: 'whatsErName',
        aString: 'More Content',
        aNumber: 789,
        bNumber: 0.43,
        cNumber: 0.1,
        aBoolean: true,
        aDatetime: 1579349481847,
        aCategorical: 34,
        aOrdinal: -1,
        aLayout: { x: 0.1134, y: 0.7356 },
        aLocation: { latitude: 13.9756, longitude: -27.443 },
      },
    ],
  },
};

describe('export data function', () => {
  it('should create valid xml for network data', () => {
    const xmlResult = ExportData(sessionA.network, variableRegistry, () => {});
    expect(xmlResult).toMatchSnapshot();
  });

  it('translates node primary key to "id" attribute', () => {
    const xml = ExportData(sessionA.network, variableRegistry, () => {});
    const doc = new DOMParser().parseFromString(xml, 'application/xml');
    const node = doc.querySelector('graph node:first-child');
    expect(node.id).toEqual(`${sessionA.network.nodes[0][NodePK]}`);
    expect(node.querySelector(`[key="${NodePK}"]`)).toBe(null);
  });
});
