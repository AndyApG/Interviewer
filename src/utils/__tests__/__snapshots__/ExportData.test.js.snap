// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export data function should create valid xml for network data 1`] = `
"<graphml xmlns=\\"http://graphml.graphdrawing.org/xmlns\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://graphml.graphdrawing.org/xmlns
http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd\\">
  <key id=\\"x\\" attr.name=\\"x\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"y\\" attr.name=\\"y\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"type\\" attr.name=\\"type\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"name\\" attr.name=\\"name\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"aString\\" attr.name=\\"aString\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"aNumber\\" attr.name=\\"aNumber\\" attr.type=\\"integer\\" for=\\"node\\"></key><key id=\\"bNumber\\" attr.name=\\"bNumber\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"cNumber\\" attr.name=\\"cNumber\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"aBoolean\\" attr.name=\\"aBoolean\\" attr.type=\\"boolean\\" for=\\"node\\"></key><key id=\\"aDatetime\\" attr.name=\\"aDatetime\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"aCategorical\\" attr.name=\\"aCategorical\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"aOrdinal\\" attr.name=\\"aOrdinal\\" attr.type=\\"integer\\" for=\\"node\\"></key><key id=\\"aLayoutX\\" attr.name=\\"aLayoutX\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"aLayoutY\\" attr.name=\\"aLayoutY\\" attr.type=\\"double\\" for=\\"node\\"></key><key id=\\"aLocation\\" attr.name=\\"aLocation\\" attr.type=\\"string\\" for=\\"node\\"></key><key id=\\"type\\" attr.name=\\"type\\" attr.type=\\"string\\" for=\\"edge\\"></key><key id=\\"connected\\" attr.name=\\"connected\\" attr.type=\\"boolean\\" for=\\"edge\\"></key><graph edgedefault=\\"undirected\\">
 <node id=\\"1\\"><data key=\\"type\\">person</data><data key=\\"name\\">soAndSo</data><data key=\\"aString\\">content</data><data key=\\"aNumber\\">123</data><data key=\\"bNumber\\">0.33</data><data key=\\"cNumber\\">3</data><data key=\\"aBoolean\\">true</data><data key=\\"aDatetime\\">1529349451847</data><data key=\\"aCategorical\\">24</data><data key=\\"aOrdinal\\">1</data><data key=\\"aLayoutX\\">0.4134</data><data key=\\"aLayoutY\\">0.2356</data><data key=\\"aLocation\\">{\\"latitude\\":41.799756,\\"longitude\\":-87.66443}</data><data key=\\"x\\">423.3216</data><data key=\\"y\\">587.0591999999999</data></node><node id=\\"2\\"><data key=\\"type\\">person</data><data key=\\"name\\">whoDunnit</data><data key=\\"aString\\">Another Content</data><data key=\\"aNumber\\">456</data><data key=\\"bNumber\\">1.8</data><data key=\\"cNumber\\">33</data><data key=\\"aBoolean\\">false</data><data key=\\"aDatetime\\">1529343481847</data><data key=\\"aCategorical\\">77</data><data key=\\"aOrdinal\\">2</data><data key=\\"aLayoutX\\">0.3434</data><data key=\\"aLayoutY\\">0.3156</data><data key=\\"aLocation\\">{\\"latitude\\":44.9756,\\"longitude\\":18.443}</data><data key=\\"x\\">351.6416</data><data key=\\"y\\">525.6192</data></node><node id=\\"3\\"><data key=\\"type\\">person</data><data key=\\"name\\">whatsErName</data><data key=\\"aString\\">More Content</data><data key=\\"aNumber\\">789</data><data key=\\"bNumber\\">0.43</data><data key=\\"cNumber\\">0.1</data><data key=\\"aBoolean\\">true</data><data key=\\"aDatetime\\">1579349481847</data><data key=\\"aCategorical\\">34</data><data key=\\"aOrdinal\\">-1</data><data key=\\"aLayoutX\\">0.1134</data><data key=\\"aLayoutY\\">0.7356</data><data key=\\"aLocation\\">{\\"latitude\\":13.9756,\\"longitude\\":-27.443}</data><data key=\\"x\\">116.1216</data><data key=\\"y\\">203.05919999999998</data></node><edge id=\\"0\\" source=\\"2\\" target=\\"1\\"><data key=\\"type\\">friend</data><data key=\\"connected\\">true</data></edge></graph>
 </graphml>"
`;
